### PR TITLE
Separated Configuration into its own class

### PR DESCRIPTION
**Pull request by drewsonne**
Added a new class to encapsulate the configuration from the business logic.
This is to allow a defaults section as well as easier testing
This is a way to allow a global "save_password" or a config specific value. In preparation for #37 

Would there be an appetite for creating a release after this in the interest of not having a release which is too large with too many changes?

### CODE REVIEW

**Comment by drewsonne** [comment]
Can I hold off on merging this PR? I will do some user testing when I get home to make sure it all still works, after the reverts and merges, etc.

**Comment by drewsonne** [comment]
Ok, I've done some more testing on this, and it looks happy and ready to be reviewed

**Comment by cameronmarlow** [comment]
This looks great - just a heads up that I might not get to a review until the weekend.

**Comment by drewsonne** [comment]
@cameronmarlow now worries! In the meantime, I may expand test cover a bit more, but that should not affect functionality

**Comment by drewsonne** [comment]
Hi @cameronmarlow have you had a chance to look at this?

**Comment by drewsonne** [comment]
Have added a smaller PR to make the delta on this PR a bit smaller

https://github.com/physera/onelogin-aws-cli/pull/44.

File: onelogin_aws_cli/__init__.py [hunk 1]
```diff
@@ -6,39 +6,13 @@
...
 import getpass
 import os
-import xml.etree.ElementTree as ET
+import xml.etree.ElementTree as ElementTree
 
 import boto3
 from onelogin.api.client import OneLoginClient
 
 CONFIG_FILENAME = ".onelogin-aws.config"
-
```
Comment (drewsonne) [inline onelogin_aws_cli/__init__.py]: When trying to build a test case, I found that keeping this function in the `__init__.py` file was causing circular dependencies with imports, so I moved it into its own module , `onelogin_aws_cli.userquery`

File: onelogin_aws_cli/__init__.py [hunk 2]
```diff
@@ -6,39 +6,13 @@
 import configparser
 import getpass
 import os
-import xml.etree.ElementTree as ET
+import xml.etree.ElementTree as ElementTree
```
Comment (drewsonne) [inline onelogin_aws_cli/__init__.py]: `flake8` was complaining about short import aliases, hence the rename.

File: onelogin_aws_cli/__init__.py [hunk 3]
```diff
@@ -221,42 +196,3 @@ def save_credentials(self):
         # Reset state in the case of another transaction
         self.token = None
         self.credentials = None
-
-    @staticmethod
```
Comment (drewsonne) [inline onelogin_aws_cli/__init__.py]: The logic for `generate_config()` is now found in `onelogin_aws_cli.ConfigurationFile.initialise()`

File: onelogin_aws_cli/configuration.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        Return a single config file section to disk
+        :param section_name:
+        :return:
+        """
+        if not self.has_section(section_name):
+            self.add_section(section_name)
+        return Section(section_name, self)
+
+
+class Section(object):
```
Comment (drewsonne) [inline onelogin_aws_cli/configuration.py]: The built-in `configparser.ConfigParser` class does not represent config file sections on their own, so I have included a small wrapper class to behave like a dictionary to reduce the amount of work needed to integrate it.

File: onelogin_aws_cli/configuration.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
+"""
+Static User Configuration models
+"""
+import configparser
+
+from onelogin_aws_cli.userquery import user_choice
+
+
```
Comment (drewsonne) [inline onelogin_aws_cli/configuration.py]: There are two reasons to abstract this out:

- encapsulation of the config class where it is aware of it's own file path, makes it easier to mock the file pointer for testing
 - Allow the definition of a `defaults` section in the config file to allow values in `[defaults]` to be applied to all sections unless over-ridden.
 - reduce the size of the `onelogin_aws_cli.OneloginAWS` class and

File: onelogin_aws_cli/__init__.py [hunk 6]
```diff
@@ -6,39 +6,13 @@
...
-        try:
-            val = int(choice) - 1
-            if val in range(0, len(options)):
-                selection = options[val]
-            else:
-                print("Invalid option")
-        except ValueError:
-            print("Invalid option")
-    return selection
+DEFAULT_CONFIG_PATH = os.path.expanduser("~/{}".format(CONFIG_FILENAME))
```
Comment (drewsonne) [inline onelogin_aws_cli/__init__.py]: I may be getting ahead of myself but this opening the door to a hierarchy of paths (including environment variables) to look for configs. This could allow for system-wide configs installed by admins, so the user config file only contains user-specific configurations.

File: onelogin_aws_cli/configuration.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    def save(self):
+        """
+        Save this config to disk
+        """
+        self.write(self.file)
+        print("Configuration written to '{}'".format(self.file))
+
+    def section(self, section_name):
+        """
+        Return a single config file section to disk
```
Comment (slycoder) [inline onelogin_aws_cli/configuration.py]: The prose in the comment is a little awkward here.

File: onelogin_aws_cli/configuration.py (hunk 4 above)
Comment (slycoder) [reply onelogin_aws_cli/configuration.py]: Neat idea!

File: onelogin_aws_cli/__init__.py [hunk 8]
```diff
@@ -6,39 +6,15 @@
...
-            val = int(choice) - 1
-            if val in range(0, len(options)):
-                selection = options[val]
-            else:
-                print("Invalid option")
-        except ValueError:
-            print("Invalid option")
-    return selection
+CONFIG_FILENAME = ".onelogin-aws.config"
+DEFAULT_CONFIG_PATH = os.path.expanduser("~/{}".format(CONFIG_FILENAME))
```
Comment (slycoder) [inline onelogin_aws_cli/__init__.py]: os.path.join?

File: onelogin_aws_cli/tests/test_configurationFile.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+[profile_test]
+save_password = false""")
+        str.seek(0)
+        cfg = ConfigurationFile(str)
+        self.assertFalse(cfg.section("profile_test").can_save_password)
+
+    def test_can_save_password_username_xor_1(self):
+        str = StringIO()
+        str.write("""[defaults]
+save_password = false
```
Comment (slycoder) [inline onelogin_aws_cli/tests/test_configurationFile.py]: Some of these seem repeated.  Did you intend for this to be true / false?

File: onelogin_aws_cli/tests/test_configurationFile.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+[profile_test]
+save_password = false""")
+        str.seek(0)
+        cfg = ConfigurationFile(str)
+        self.assertFalse(cfg.section("profile_test").can_save_password)
+
+    def test_can_save_password_username_xor_3(self):
```
Comment (slycoder) [inline onelogin_aws_cli/tests/test_configurationFile.py]: For my OCD, could you add a test for defaults = True and empty elsewhere?

File: onelogin_aws_cli/__init__.py (hunk 8 above)
Comment (drewsonne) [reply onelogin_aws_cli/__init__.py]: Right now, we only store in the home directory, and it's about making sure this still works on windows (I hope). If I use os.path.join, I still have to go find the home dir.

https://stackoverflow.com/questions/4028904/how-to-get-the-home-directory-in-python

File: onelogin_aws_cli/tests/test_configurationFile.py (hunk 9 above)
Comment (drewsonne) [reply onelogin_aws_cli/tests/test_configurationFile.py]: These were written before I split the PR for the save-username functionality, and there may be duplicates. Let me have a closer look

File: onelogin_aws_cli/__init__.py (hunk 8 above)
Comment (drewsonne) [reply onelogin_aws_cli/__init__.py]: Just to clarify, were you suggesting I use `os.path.join`?

File: onelogin_aws_cli/tests/test_configurationFile.py (hunk 10 above)
Comment (drewsonne) [reply onelogin_aws_cli/tests/test_configurationFile.py]: Done,

File: onelogin_aws_cli/__init__.py (hunk 8 above)
Comment (slycoder) [reply onelogin_aws_cli/__init__.py]: Yeah, I think it'd be safer on windows to do
os.path.join(os.path.expanduser("~"), CONFIG_FILENAME)

File: onelogin_aws_cli/tests/test_configurationFile.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+    def _helper_build_config(self, config_content: str):
+        str = StringIO()
+        str.write(config_content)
+        str.seek(0)
+        return ConfigurationFile(str)
+
+    def test_can_save_password_username_true(self):
+        cfg = self._helper_build_config("""[defaults]
+save_password = false
```
Comment (slycoder) [inline onelogin_aws_cli/tests/test_configurationFile.py]: This is still duplicated with xor1.  If you make this line true, you'll have covered all cases I believe =).

**Comment by slycoder** [comment]
Thanks for all the great contributions @drewsonne .  Would you be interested in being a maintainer of the repo while you're at it ? 👍

**Comment by drewsonne** [comment]
@slycoder I would definitely be interested in this. I appreciate you guys have a business to run and this is a side project (sort-of), so I would definitely be interested in helping out. 

I will continue to expand the test coverage, as I also will be using this in a corporate setting. That said, I hope there is still the possibility of reviews and such, as I myself don't trust my own code. :-)

But yes, definitely interested in being a Maintainer.
